### PR TITLE
feat: support custom package, headers and body in openwhisk plugin

### LINKED ISSUE #5756: request help: OpenWhisk plugin does not support modifying the response
## Description
The current `openwhisk` plugin will take the response from Apache OpenWhisk and return it directly without parsing. This request may contain complex data that rewrites the status code, response headers, and response body. We need to parse it and change the APISIX response.

The complex data like this:
```text
{
  "statusCode": 401,
  "headers": {
    "addition-header": "a-header"
  },
  "body": "xxx" | {"xxx":"xxx"}
}
```
We need to determine if the OpenWhisk response contains them, and if they are present, write them to the APISIX response.

### PR DESCRIPTION
**Pull request by encanto-oss**
### What this PR does / why we need it:

This PR added openwhisk response parsing mechanism, implementing issue #5756. As of now, user defined openwhisk package action, status code and body are supported.

Fix #5756

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/openwhisk.lua [hunk 1]
```diff
@@ -104,8 +106,35 @@ function _M.access(conf, ctx)
         return 503
     end
 
+    -- parse OpenWhisk JSON response
+    -- OpenWhisk supports two types of responses, the user can return only
+    -- the response body, or set the status code and header.
+    local result, err = core.json.decode(res.body)
+
+    if err then
```
Comment (spacewander) [inline apisix/plugins/openwhisk.lua:114]:
```suggestion
    if not result then
```
would be more common

File: apisix/plugins/openwhisk.lua [hunk 2]
```diff
@@ -104,8 +106,35 @@ function _M.access(conf, ctx)
...
     -- setting response headers
-    core.response.set_header(res.headers)
+    if result.headers ~= nil then
+        core.response.set_header(result.headers)
+    end
+
+    -- return status code and body
+    if result.statusCode ~= nil and result.body ~= nil then
+        return result.statusCode, result.body
+    end
```
Comment (spacewander) [inline apisix/plugins/openwhisk.lua:127]:
```
if result.statusCode ~= nil then
    return result.statusCode, result.body
end
```
already covers two cases?

File: apisix/plugins/openwhisk.lua [hunk 3]
```diff
@@ -30,6 +30,7 @@ local schema = {
         },
         service_token = {type = "string"},
         namespace = {type = "string", maxLength = 256},
+        package = {type = "string", maxLength = 256},
```
Comment (spacewander) [inline apisix/plugins/openwhisk.lua:33]: Do we need to add a regex pattern to validate the input, like not `/` can be used in the package name?

File: apisix/plugins/openwhisk.lua (hunk 1 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua:114]: fixed.

File: apisix/plugins/openwhisk.lua (hunk 2 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua:127]: changed.

File: apisix/plugins/openwhisk.lua (hunk 3 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua]: added regex validation according to [https://github.com/apache/openwhisk/blob/master/docs/reference.md#entity-names](https://github.com/apache/openwhisk/blob/master/docs/reference.md#entity-names)

File: apisix/plugins/openwhisk.lua (hunk 2 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua]: The statuscode and body may not exist simultaneously, we need to handle different cases respectively.

Comment (spacewander) [reply apisix/plugins/openwhisk.lua]: Return a `nil` `result.body` is equal to `return result.statusCode`. So we can save one branch.

File: apisix/plugins/openwhisk.lua [hunk 4]
```diff
@@ -104,8 +108,31 @@ function _M.access(conf, ctx)
         return 503
     end
 
+    -- parse OpenWhisk JSON response
+    -- OpenWhisk supports two types of responses, the user can return only
+    -- the response body, or set the status code and header.
+    local result, err = core.json.decode(res.body)
```
Comment (tzssangglass) [inline apisix/plugins/openwhisk.lua:114]: should check the `res.body` is not nil?

File: apisix/plugins/openwhisk.lua [hunk 5]
```diff
@@ -104,8 +108,31 @@ function _M.access(conf, ctx)
...
+        return 503
+    end
+
     -- setting response headers
-    core.response.set_header(res.headers)
+    if result.headers ~= nil then
+        core.response.set_header(result.headers)
+    end
+
+    if result.statusCode ~= nil and result.body ~= nil then
```
Comment (tzssangglass) [inline apisix/plugins/openwhisk.lua:126]: we can use `local code = result.statusCode` and `local body = result.body`, and then use `code` and `body`

File: apisix/plugins/openwhisk.lua [hunk 6]
```diff
@@ -104,8 +108,31 @@ function _M.access(conf, ctx)
...
+        return result.statusCode, result.body
+    elseif result.statusCode ~= nil then
+        -- return only status code
+        return result.statusCode
+    elseif result.body ~= nil then
+        -- return only body
+        return 200, result.body
+    end
 
     return res.status, res.body
```
Comment (shuaijinchao) [inline apisix/plugins/openwhisk.lua:137]: this should be better.
```lua
local code = result.statusCode or res.status
local body = result.body or res.body
return code, body
```

File: apisix/plugins/openwhisk.lua (hunk 5 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua:126]: resolved.

File: apisix/plugins/openwhisk.lua (hunk 6 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua:137]: changed.

File: apisix/plugins/openwhisk.lua (hunk 2 above)
Comment (encanto-oss) [reply apisix/plugins/openwhisk.lua]: Applied @shuaijinchao suggestion.

**Review by tzssangglass — APPROVED** [review]

File: apisix/plugins/openwhisk.lua [hunk 7]
```diff
@@ -104,10 +108,30 @@ function _M.access(conf, ctx)
...
+    if res.body == nil then
+        return res.status, res.body
+    end
+
+    -- parse OpenWhisk JSON response
+    -- OpenWhisk supports two types of responses, the user can return only
+    -- the response body, or set the status code and header.
+    local result, err = core.json.decode(res.body)
+
+    if not result or err then
```
Comment (spacewander) [inline apisix/plugins/openwhisk.lua:121]:
```suggestion
    if not result then
```

**Review by spacewander — APPROVED** [review]

**Pull request merged**